contracts/Keyboards.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+contract Keyboards {
+  enum KeyboardKind { 
+    SixtyPercent, 
+    SeventyFivePercent, 
+    EightyPercent, 
+    Iso105 
+  }
+
+  struct Keyboard {
+    KeyboardKind kind; 
+    // ABS = false, PBT = true
+    bool isPBT;
+    // tailwind filters to layer over
+    string filter;
+    // user who created it! 
+    address owner;
+  }
+
+  Keyboard[] public createdKeyboards;
+
+  event KeyboardCreated(
+    Keyboard keyboard
+  );
+
+  function create(
+    KeyboardKind _kind,
+    bool _isPBT,
+    string calldata _filter
+  ) external {
+    Keyboard memory newKeyboard = Keyboard({
+      kind: _kind,
+      isPBT: _isPBT,
+      filter: _filter,
+      owner: msg.sender
+    });
+
+    createdKeyboards.push(newKeyboard);
+    emit KeyboardCreated(newKeyboard);
+  }
+
+  function tip(address payable _owner) external payable  {
```

Comment:
Ok, quite curious about the edge cases around this function. It looks like anyone could call it with the current access modifiers. What would happen if some rando tried to call this, could they tip anyone with money stored in this contract?  (also i believe this is the correct way to think about it, but the account that the money would be transferred from is the address of this deployed contract and it would basically act as its own wallet?)

Reply:
Anyone can call it, but they can't control the `msg.value` in `_owner.transfer(msg.value);` freely - that's always the amount of eth transferred in their call to the function. 

So what this function does is transfer whatever the user pays in the transaction where they call this function, to the given address. It's not strictly necessary because the user could just transfer it directly on the client without calling the contract, so it's more just to demonstrate how to send payments.

You're right that the contract **can** have a balance, and that balance can be paid out by it using a function like this. But in this case the balance of the contract is always 0, because any money paid to it in this function (which is the only payable one) is immediately paid out to the `_owner` address.

Reply:
ohhh, ok i think i get ittt. so then humor me, would this be a dangerous function??
```
function tip(address payable _owner, uint256 value) external {
  _owner.transfer(value);
}
```
because the function is no longer payable and the amount just comes through as an argument?

Reply:
Yep, that would allow any user to make our contract pay any amount to any address!

That's definitely a bad idea :p Whether it's dangerous kinda depends on the contract though. Basically that function will only work if the contract balance is >= `value`, otherwise it'll fail. There's two ways (that I know of!) that a contract's balance can increase: you can deploy it with an initial balance, or you can have a `payable` function. 

So if you swapped our `tip` function out for that one, and then you deployed it with a `0` balance, then that function wouldn't really be dangerous because it would just fail to ever pay out any `value >= 0`. But if the contract had some other payable function, or we deployed it with a balance that we wanted to have managed by the contract, then that function would allow any caller to steal the whole balance.

Reply:
sweet, that definitely helped my mental model, thanks!